### ISSUE TITLE
Magic numbers for metric overrides to render nicely?

### ISSUE THREAD

**Issue by pallaswept** [be5invis/Iosevka#2441]
Hi belleve invis :)

Thanks once again for this amazing font. It's really been helpful for me. 

This is not really a FR or a bug, more of a request for advice. I understand that I am discussing unsupported features, and I have zero expectation of assistance, but any that you might be able to offer, would be greatly appreciated!

### TL;DR in case you are busy
in order to avoid ugly rendering of our custom built Iosevka, are there some magic numbers we need to know, when using metric overrides?
I especially have trouble with modifying side bearings (sb=) or widths. Any change to the characters' horizontal measurements, seems to give bad results. sb is the big one for me.

### Why I ask:
I became curious about some of the custom build features, and thought that maybe I might make some 'personal taste' tweaks to the font, and just enjoy experimenting with it. Several things I've tinkered with, are in the metric overrides section. 

Mostly, I found that I am an idiot and know nothing about fonts :laughing: but it is a learning experience, and while sometimes the results were bad, and often I just made it crash, but some of my results have been very satisfying.

I did have quite some trouble with getting the thing to behave nicely when rendered, though. Changes to things internal to the character like archdepth didn't cause me any trouble, but when I changed side bearings... I broke it.  Normally, I use full hinting (in my fontconfig settings via KDE) to try to minimize fuzzy rendering, so the letters snap to pixels quite well, but I did notice that when I tried my custom Iosevka build with slight hinting, the results were....honestly, terrible. 

Especially noticeable at smaller font sizes where the stroke weight was in-between 1 and 2 pixels wide, some letters were too wide and some too narrow, some seemed heavier than others, etc. This was the result of using, for example, `sb = 'default_sb / 1.2'` (When using metric overrides, I try to use some multiple of the default_* number, so that it retains as much of the stock 'magic' as possible). 

Similarly, I had trouble when using custom widths (eg 518 instead of 500). The changing horizontal position of the strokes seemed to make them not align with the pixel grid very well. The result was the same, 'lumpy' fonts where they are too thick or thin, dark or light, etc.

Here is an example of such a failure. This one I compiled by accident, but it's a good example. 
This is based on Iosevka Aile, changes here were `sb='default_sb / 1.2'` and width 460.: 
![image](https://github.com/user-attachments/assets/f3c0a8e1-9678-4aba-9066-8cba85f06f14)
(sorry for the weird scaling effect, it's KDE being weird about how it zooms in)

Note the different rendering of the two 'o' characters here. It's quite unpleasant to look at, in person, it 'jumps out' at me. With stock Iosevka I don't see behaviour like this. 
This one I compiled by accident, but it's a good type-example of the changes I'm discussing and their effects. This is based on Iosevka Aile, changes here were `sb='default_sb / 1.2'` and width 460. I am using full hinting in the above screenshot - if I use partial hinting, the distortions are different, but still present, and usually visually worse.

I also noticed a few strange distortions when altering xheight - for example, I would see in ligatures like --- or -->, the middle hyphen would be at a different height. I could see it in KDE's font preview when I zoomed in and out. At some zoom levels, it would be correct, at others, not. Note how in these screenshots, as I scroll the mouse wheel to change font size, the hyphen (-) is sometimes above and sometimes below the horizontal bar in the plus (+). But I look at the source code and it seems (to my untrained eye) as though the exact same code calculates these two strokes - so how could they appear differently? I must be doing something noob.
Also, it is more subtle, but I notice that in the plus character, the vertical stroke is not centred on the horizontal stroke - sometimes there is more above, sometimes more below that stroke.

![image](https://github.com/user-attachments/assets/7a62bdb5-79de-4da1-a04b-07456ba7fc1f)
![image](https://github.com/user-attachments/assets/d9c47ea2-ca30-46cc-9941-89176887b7bd)
![image](https://github.com/user-attachments/assets/0188af51-7dcd-4a4c-8cf8-098da97a8d67)

### Are there 'magic' numbers?
I did notice in the source code that a lot of the parameters for metrics are not 'nice', 'round' numbers, or even weird-looking decimals that represent simple fractions... they look very deliberately chosen, as if maybe they are the results of calculations, or formula you apply, or something? I feel like perhaps the 'intuitive' numbers I am using are just not the right kind of thing, because the ones you use are not so 'nice' like 1.2 aka 6/5, they are usually more complex like 1.203 or something. 

So... Is there a magic trick to selecting numbers that will render well? If the answer is "yes, but it's too complicated to teach you in github comments" then I totally understand. Perhaps this might be a thing where you could drop me some useful search engine terms to go and read and learn some more... Or maybe you will just tell me "always use prime numbers" or "use numbers divisible by a grid this size" or something easy :D

Thanks in advance for any advice you might offer.

**Comment by be5invis** [comment]
When designing glyphs I look at them at large sizes first. There are fractions like the 0.525 and 0.09 in H's middle bar placement (_cap height_ × 0.525 − 0.09 × _stroke width_)? Why 0.525 and 0.09? Well, I build the font for many times with different parameters and choose the one that looks the most pleasant...

Iosevka currently don't have any specialized design for snapping to grids — But for your Konsole case, it might be related to "subpixel placement", i.e., allowing a glyph to have a width that is not an integer pixel. Perhaps you can adjust your fontconfig settings to use whole-pixel placement only for Iosevka.

**Comment by pallaswept** [comment]
OK, so the "magic" numbers are not some mysterious equation, but are a reflection of your vast experience - it's 'art'. Well, no wonder it looks so good :)

... and the rendering problems I'm seeing are problems with the font rendering tools rather than my font... 

I think we can close this one now :) Thank you very much for taking the time to explain it to me. Have a great week!

**Comment by pallaswept** [comment]
I was going to edit the above post so to not spam notifications, I was just leaving an update for future travellers who may have similar problems (with any font!)... But perhaps my findings might be useful info for you to pass on to other users if they have similar problems, so I will post it here.

**TL;DR:** I do not recommend the use of 'Full hinting' with greyscale AA. 
The fontconfig issues I was seeing are the result of Full hinting. Switching to slight hinting has resulted in better shapes, as one would expect, but, with greyscale AA, at 1080p, has not significantly increased blurring. At higher resolutions this effect will be even more pronounced. I confirmed this at 1440p and 4K.

Long version:
**Which settings for Iosevka?**
I have often wondered which was most appropriate for Iosevka, since it has a whole hinting process for the fonts, I figured one setting might be preferable to the other. The fontconfig/freetype documentation does not explain this well at all - it says that "Slight hinting invokes the native hinter if possible" but does not exactly explain (at least, not in a way understandable by new people like me!) what the "native" hinter is (native to what?), or what defines whether it is "possible".

It does say of slight hinting, "This mode implied the auto-hinter before and has now been changed to mean “Use native vertical-grid-only-snapping if driver and font supports it and vertical-grid-only auto-hinter otherwise”. It also suggests that perhaps "native" means "native to the font" not "native to the system" "native to the DE" or something else: "In the past, setting ‘slight’ hinting via FontConfig or configuration GUIs meant that native hints within a font were ignored"

**How to get sharper rendering?**
So anyway, since I'm trying for sharper rendering, most documentation (for freetype, fontconfig, or just in general) says the secret is good pixel alignment, and that Full hinting is the way (they also usually suggest subpixel AA which one should read as an instant red flag that this doco is old and to be largely ignored!). But it is expected that it might cause some distortion of the font, and it just was pure bad luck that my modifications to Iosevka exposed said distortion in an extreme fashion. Having re-read the above doc, I concluded that I should try going back to Slight hinting, and the results are an improvement on the above:

![image](https://github.com/user-attachments/assets/fc308cea-615d-4a13-99ba-4c127c768b9c)

**Outdated docs are a thing**
This is to be expected, less hinting would normally generate a better character shape, but, at the cost of more blurred edges due to the need for more antialiasing. And this is what all those pages advise us when we google/duck/bing/brave search it. But I think they are a good example of out-dated (and incomplete) info online. Back when most of these articles were written, the standard was low DPI displays and subpixel AA. Years ago, however, the standard was improved by Apple, to high DPI ('retina' for them) and greyscale AA. In addition to reasons of modernity, greyscale AA has always been a requirement for multiple displays with heterogenous subpixel layouts (you can't use subpixel rendering if you have one RGB display and one BGR or one RGB but mounted vertically)

**Findings**
Using this more modern (and to me and many others, less eye-straining) approach of greyscale AA, implies more antialiased pixels always, because we essentially have 1/3 the number of pixels now, so the same thing that used to require one antialiased subpixel, now requires three subpixels/one whole pixel. This means that even with pixels snapped to the grid where possible, where it is not possible to align to the pixels exactly (almost always because strokes aren't perfectly pixel-width) there will be AA pixels, and whatever AA there is, will be triple compared to subpixel AA. So, the reduced AA from hinting/pixel-snapping, is less of a benefit now. Yes, it is sharper than less-hinted approaches, but not *that* much sharper.

**Conclusions**
Accordingly, if the benefit of more hinting (less AA) is reduced, the cost of less hinting (more AA) is now relatively lower, while the cost of more hinting is still high (distorted characters as above) so it hurts as much as before... So I'd say the modern (in the 2020's) advice for a person using greyscale AA, is to just use Slight hinting. If you want sharper rendering, buy a better screen, it's simply not gonna happen in software. Subpixel AA _might_ make it sharper (maybe, depending on your display), but with other costs like rainbow fringes and not working with modern software which expects greyscale AA, plus several rather severe Windows-specific problems which are another rabbit-hole.... 

So in conclusion: don't use subpixel AA in 2024, and **if you're using greyscale AA, don't use full hinting** - Use greyscale AA and use Slight hinting, and it will use Iosevka's built-in hints (and employ the auto-hinter for fonts which lack them). The shape will be 100x better and the blur just barely worse if at all.

Hope this helps some future traveller one day.
